Use passive event listener for touchmove (to prevent scrolling while swiping)

xanido opened the pull request:
This small patch restores react-slider's ability to disable vertical scrolling when swiping through a slider on iOS.

Seems the `touchmove` event listener that React adds (when using the React's event system via event props) is not configured as a non-passive listener. iOS ([starting with iOS 11.3](https://stackoverflow.com/a/49582193)) has made all _root document_ touch event listeners passive by default to improve scrolling performance. React's event system attaches event handlers to the document root, which will result in a passive `touchmove` listener on iOS, which in turn means that `preventDefault()` has no effect.

The proposed solution in https://github.com/facebook/react/issues/2043 should fix the issue, but for now adding the event listener directly to the node also solves the problem.

Please note, the `{passive: false}` in the addEventListener call in this PR is not technically necessary as only _document root_ event listeners are passive by default on iOS, but it does explicitly communicate the intent.

xanido committed:
Use passive event listener for touchmove

mattcolman commented on src/inner-slider.js:
```diff
@@ -188,6 +195,13 @@ export class InnerSlider extends React.Component {
     this.adaptHeight();
+
+    if (prevProps.touchMove !== this.props.touchMove) {
+      const method = touchMove ? "add" : "remove";
```
is touchMove defined? Should be this.props.touchMove?

xanido replied:
Sorry, fixed!

mattcolman commented on src/inner-slider.js:
```diff
@@ -704,7 +714,6 @@ export class InnerSlider extends React.Component {
       onMouseUp: touchMove ? this.swipeEnd : null,
       onMouseLeave: this.state.dragging && touchMove ? this.swipeEnd : null,
       onTouchStart: touchMove ? this.swipeStart : null,
-      onTouchMove: this.state.dragging && touchMove ? this.swipeMove : null,
```
do you also need to preserve this logic around `this.state.dragging`?

xanido replied:
I've added it back in but in a slightly different manner to work around another iOS Safari quirk (touch listeners registered by other touch listeners don't seem to be able to be made non-passive).

mattcolman commented on src/inner-slider.js:
```diff
@@ -97,6 +97,11 @@ export class InnerSlider extends React.Component {
     );
+
+    if (this.props.touchMove) {
+      this.list.addEventListener("touchmove", this.swipeMove);
```
did you remove { passive: false } because it's the default? I'd probably keep it and add a comment to explain that we need passive set to false to allow preventDefault and link to https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener for more details.

xanido replied:
I felt that it was unnecessary to leave it in, as false is the default. I can add it back in to guard against a future change to the default (although I'd guess it'd be very unlikely to happen)

xanido replied:
Added it back in and added note

xanido committed:
Make touchmove listener non-passive explicitly

xanido committed:
Refactor 'is dragging' logic for iOS

Previously touchmove handler was registered from within a touchstart event.
Seems as though handlers registered from within touch handlers are
always passive, regardless of the passive param. Instead we keep the touchmove
handler around, and return early if not in a dragging state.

mattcolman reviewed (approved):
looks good to me! Awesome work @xanido !

mattcolman commented:
oh one last thing, might need to clean up touchmove in `componentWillUnmount`?

xanido committed:
Cleanup touchmove handler on unmount

jooj123 commented on src/inner-slider.js:
```diff
@@ -188,6 +202,15 @@ export class InnerSlider extends React.Component {
+      const method = this.props.touchMove ? "add" : "remove";
+      // swipeMove calls preventDefault which only has an effect on non-passive events
+      // https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener
+      this.list[`${method}EventListener`]("touchmove", this.swipeMove, {
```
isnt there a chance that there is a listener thats created but not removed here?

xanido commented:
D’oh you’re right! Sorry thanks for catching that, will update

On Fri, 31 May 2019 at 1:33 pm, Matt Colman <notifications@github.com>
wrote:

> *@mattcolman* commented on this pull request.
> ------------------------------
>
> In src/inner-slider.js
> <https://github.com/DomainGroupOSS/react-slick/pull/2#discussion_r289247375>
> :
>
> > @@ -188,6 +195,13 @@ export class InnerSlider extends React.Component {
>      //   this.props.onLazyLoad([leftMostSlide])
>      // }
>      this.adaptHeight();
> +
> +    if (prevProps.touchMove !== this.props.touchMove) {
> +      const method = touchMove ? "add" : "remove";
>
> is touchMove defined? Should be this.props.touchMove?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DomainGroupOSS/react-slick/pull/2?email_source=notifications&email_token=AAGUQWFSDKA6T3BQDOBZ2J3PYCMBLA5CNFSM4HRQOHTKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2GP4WI#pullrequestreview-244121177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGUQWCMC4B4NC3T3QTTC2LPYCMBLANCNFSM4HRQOHTA>
> .
>

jooj123 commented on src/inner-slider.js:
```diff
@@ -118,6 +127,11 @@ export class InnerSlider extends React.Component {
     }
+    if(this.props.touchMove) {
+      this.list.removeEventListener("touchmove", this.swipeMove, {
+        passive: false
```
good to check if `animating` is true, if its true you can have passive true which will improve performance
as you are promising the handler won't call `preventDefault` to disable scrolling.

jooj123 commented on src/inner-slider.js:
```diff
@@ -97,6 +97,15 @@ export class InnerSlider extends React.Component {
+      // swipeMove calls preventDefault which only has an effect on non-passive events
+      // https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener
+      this.list.addEventListener("touchmove", this.swipeMove, {
+        passive: false
```
good to check if `animating` is true, if its true you can have passive true which will improve performance
as you are promising the handler won't call `preventDefault` to disable scrolling.

jooj123 commented:
why do you need to disable vertical scrolling when swiping ?

mattcolman commented:
That's the expected behaviour, if you are trying to swipe horizontal it won't also allow you to scroll vertically.
